tests/migrations/test_operations.py

```diff
@@ -4058,6 +4058,68 @@ def test_alter_fk_non_fk(self):
         self.assertColumnExists("test_afknfk_rider", "pony_id")
         self.assertColumnNotExists("test_afknfk_rider", "pony")
 
+    def test_multiple_alter_fields_data_with_default_value(self):
+        """
+        Altering CharField to null=False with existing data and related fields (#25105)
+        """
+        app_label = "test_mafdwdv"
+
+        def create_objs(models, schema_editor):
+            Pony = models.get_model(app_label, "Pony")
+            pony = Pony.objects.create()
+            Rider = models.get_model(app_label, "Rider")
+            Rider.objects.create(pony=pony)
+
+        operations = [
+            migrations.CreateModel(
+                "Pony",
+                [
+                    ("id", models.AutoField(primary_key=True)),
+                ],
+            ),
+            migrations.CreateModel(
+                "Rider",
+                [
+                    ("id", models.AutoField(primary_key=True)),
+                    ("pony", models.ForeignKey("Pony", models.CASCADE)),
+                    (
+                        "first_name",
+                        models.CharField(max_length=50, null=True),
+                    ),
+                    (
+                        "last_name",
+                        models.CharField(max_length=50, null=True),
+                    ),
+                ],
+            ),
+            migrations.RunPython(create_objs),
+        ]
+        project_state = self.apply_operations(app_label, ProjectState(), operations)
+        self.assertColumnNull(app_label + "_rider", "first_name")
+        self.assertColumnNull(app_label + "_rider", "last_name")
+
+        # Not Null.
+        operations = [
+            migrations.AlterField(
+                "Rider",
+                "first_name",
+                models.CharField(max_length=50, null=False, default=""),
+            ),
+            migrations.AlterField(
+                "Rider",
+                "last_name",
+                models.CharField(max_length=50, null=False, default=""),
+            ),
+        ]
+        new_state = self.apply_operations(app_label, project_state, operations)
+
+        Rider = new_state.apps.get_model(app_label, "Rider")
+        rider = Rider.objects.get()
+        self.assertEqual(rider.first_name, "")
+        self.assertEqual(rider.last_name, "")
+        self.assertColumnNotNull(app_label + "_rider", "first_name")
+        self.assertColumnNotNull(app_label + "_rider", "last_name")
```

Comment:
Can we move this to the `schema` tests?, e.g.
```diff
diff --git a/tests/schema/tests.py b/tests/schema/tests.py
index 2d12796cbd..37ea0330c8 100644
--- a/tests/schema/tests.py
+++ b/tests/schema/tests.py
@@ -973,6 +973,33 @@ class SchemaTests(TransactionTestCase):
             Node.objects.update(parent=parent)
             editor.alter_field(Node, old_field, new_field, strict=True)
 
+    @isolate_apps("schema")
+    def test_alter_null_with_default_value_deferred_constraints(self):
+        class Publisher(Model):
+            class Meta:
+                app_label = "schema"
+
+        class Article(Model):
+            publisher = ForeignKey(Publisher, CASCADE)
+            title = CharField(max_length=50, null=True)
+
+            class Meta:
+                app_label = "schema"
+
+        with connection.schema_editor() as editor:
+            editor.create_model(Publisher)
+            editor.create_model(Article)
+        self.isolated_local_models = [Publisher, Article]
+
+        publisher = Publisher.objects.create()
+        Article.objects.create(publisher=publisher)
+        old_field = Article._meta.get_field("title")
+        new_field = CharField(max_length=50, null=False, default="")
+        new_field.set_attributes_from_name('title')
+        with connection.schema_editor() as editor:
+            Article.objects.update(title=None)
+            editor.alter_field(Article, old_field, new_field, strict=True)
+
     def test_alter_text_field_to_date_field(self):
         """
         #25002 - Test conversion of text field to date field.

```

Reply:
Sure :+1: 
I'm never quite sure yet where the most relevant place for tests is :)

Reply:
Fix is in `schema`, so `schema` tests seem to be a better fit.

Reply:
Of course, makes sense :sweat_smile: 
Thanks for the review and merge :ship: